### PR TITLE
Add support for block inclusion proofs

### PR DESCRIPTION
**Pull request by fiaxh**

### CODE REVIEW

File: blockchain/src/blockchain/verify.rs [hunk 1]
```diff
@@ -254,4 +276,15 @@ impl Blockchain {
 
         Ok(None)
     }
+
+    pub fn next_interlink(&self) -> Vec<Blake2bHash> {
```
Comment (sisou) [inline blockchain/src/blockchain/verify.rs]: Is this method unused?

Comment (fiaxh) [reply blockchain/src/blockchain/verify.rs]: Indeed it was, thanks.

File: consensus/src/messages/mod.rs [hunk 2]
```diff
@@ -248,3 +250,22 @@ pub struct ResponseTransactionReceiptsByAddress {
...
+#[derive(Serialize, Deserialize)]
+pub struct ResponseBlocksProof {
+    pub proof: Option<BlockInclusionProof>,
+}
+
+impl RequestCommon for RequestBlocksProof {
+    type Kind = RequestMarker;
+    const TYPE_ID: u16 = 215;
+    type Response = ResponseBlocksProof;
+    const MAX_REQUESTS: u32 = MAX_REQUEST_TRANSACTIONS_PROOF;
```
Comment (sisou) [inline consensus/src/messages/mod.rs]: Are we sure we are going to reuse the `MAX_REQUEST_TRANSACTIONS_PROOF` number for this message type?

**Review by sisou — APPROVED** [review]

File: consensus/src/messages/mod.rs (hunk 2 above)
Comment (fiaxh) [reply consensus/src/messages/mod.rs]: No, we should use the MAX_REQUEST_BLOCKS_PROOF number that I already created for that purpose, but failed to actually use :D

File: consensus/src/messages/handlers.rs [hunk 3]
```diff
@@ -6,12 +6,12 @@ use std::{cmp, sync::Arc};
 use parking_lot::RwLock;
 
 use nimiq_block::Block;
+use nimiq_block::BlockInclusionProof;
 #[cfg(feature = "full")]
 use nimiq_blockchain::{Blockchain, CHUNK_SIZE};
 use nimiq_blockchain_interface::{AbstractBlockchain, Direction};
 use nimiq_blockchain_proxy::BlockchainProxy;
 use nimiq_network_interface::request::Handle;
-#[cfg(feature = "full")]
```
Comment (jsdanielh) [inline consensus/src/messages/handlers.rs]: I think we need to leave this as it is and then add:
`#[cfg(feature = "full")]` in the handle below

Comment (jsdanielh) [reply consensus/src/messages/handlers.rs]: Fixed

File: primitives/block/src/macro_block.rs [hunk 4]
```diff
@@ -78,6 +78,38 @@ impl MacroBlock {
         Ok(pk_tree_construct(public_keys).to_vec())
     }
 
+    // Computes the next interlink from self.header.interlink
+    pub fn get_next_interlink(&self) -> Result<Vec<Blake2bHash>, BlockError> {
```
Comment (sisou) [inline primitives/block/src/macro_block.rs]:
```suggestion
    /// Computes the next interlink from self.header.interlink
    pub fn get_next_interlink(&self) -> Result<Vec<Blake2bHash>, BlockError> {
```

Comment (jsdanielh) [reply primitives/block/src/macro_block.rs]: Fixed as well

File: primitives/block/src/block_proof.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        for block in self.proof.clone() {
+            number_to_block.insert(block.block_number(), block);
+        }
+
+        let hops = BlockInclusionProof::get_interlink_hops(
+            target.block_number(),
+            election_head.block_number(),
+        );
+
+        // If we're at the 1th or 2nd macro block, there are no block proofs
```
Comment (jsdanielh) [inline primitives/block/src/block_proof.rs]:
```suggestion
        // If we're at the 1st or 2nd macro block, there are no block proofs
```

Comment (jsdanielh) [reply primitives/block/src/block_proof.rs]: Fixed

**Pull request merged**